### PR TITLE
Additional channel/chat lifecycle events

### LINKED ISSUE #2367: Additional channel/chat lifecycle events (JS)
See [parent](https://github.com/microsoft/botframework-sdk/issues/5892)

### PR DESCRIPTION
**Pull request by denscollo**
Fixes #2367

## Description
- Add the following events to the [teamsActivityHandler](https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder/src/teamsActivityHandler.ts) class.
  - teamArchived
  - teamDeleted
  - teamHardDeleted
  - teamRestored
  - teamUnarchived

**Note:** These changes can't be tested using MS-Teams until the implementation to use these events is done.

## Specific Changes
- Updated [teamsActivityHandler](https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder/src/teamsActivityHandler.ts) class to add the following methods:
  - _onTeamsTeamArchivedEvent_.
  - _onTeamsTeamDeletedEvent_.
  - _onTeamsTeamHardDeletedEvent_.
  - _onTeamsTeamRestoredEvent_.
  - _onTeamsTeamUnarchivedEvent_.

- Updated [teamsActivityHandler.tests](https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder/tests/teamsActivityHandler.test.js) class to add the unit tests for the previous methods.

## Testing
The next image shows the new tests passing:
image

### CODE REVIEW

File: libraries/botbuilder/src/teamsActivityHandler.ts [hunk 1]
```diff
@@ -460,14 +475,44 @@ export class TeamsActivityHandler extends ActivityHandler {
         await this.handle(context, 'TeamsChannelDeleted', this.defaultNextEvent(context));
     }
 
-    /**
+        /**
```
Comment (stevengum) [inline libraries/botbuilder/src/teamsActivityHandler.ts:478]: Please remove extra whitespace

File: libraries/botbuilder/src/teamsActivityHandler.ts [hunk 2]
```diff
@@ -460,14 +475,44 @@ export class TeamsActivityHandler extends ActivityHandler {
...
      */
     protected async onTeamsChannelRenamed(context): Promise<void> {
         await this.handle(context, 'TeamsChannelRenamed', this.defaultNextEvent(context));
     }
 
+    /**
+     * Invoked when a Team Archived event activity is received from the connector.
+     * Team Archived correspond to the user archiving a team.
+     * @param context The context for this turn.
+     * @returns A task that represents the work queued to execute.
```
Comment (stevengum) [inline libraries/botbuilder/src/teamsActivityHandler.ts:490]: Please fix the docstring for `@returns`

**Comment by denscollo** [comment]
@stevengum We just applied the feedback and merged with master

File: libraries/botbuilder/tests/teamsActivityHandler.test.js [hunk 3]
```diff
@@ -1526,6 +1526,221 @@ describe('TeamsActivityHandler', () => {
...
+            const adapter = new TestAdapter(async context => {
+                await bot.run(context);
+            });
+
+            adapter.send(activity)
+                .then(() => {
+                    assert(onTeamsTeamArchivedEventCalled, 'onTeamsTeamArchivedEvent handler not called');
+                    assert(onConversationUpdateCalled, 'onConversationUpdate handler not called');
+                    assert(onDialogCalled, 'onDialog handler not called');
+                });
```
Comment (stevengum) [inline libraries/botbuilder/tests/teamsActivityHandler.test.js:1569]: Are any errors from the TeamsActivityHandler, TestAdapter and TestAdapter.send().then() being properly caught and surfaced?

File: libraries/botbuilder/src/teamsActivityHandler.ts [hunk 4]
```diff
@@ -541,4 +642,28 @@ export class TeamsActivityHandler extends ActivityHandler {
...
+     * @param handler 
+     * @returns A promise that represents the work queued.
+     */
+    public onTeamsTeamUnarchivedEvent(handler: (teamInfo: TeamInfo, context: TurnContext, next: () => Promise<void>) => Promise<void>): this {
+        return this.on('TeamsTeamUnarchived', async (context, next) => {
+            const teamsChannelData = context.activity.channelData as TeamsChannelData;
+            await handler(teamsChannelData.team, context, next);
+        });
+    }
 }
```
Comment (stevengum) [inline libraries/botbuilder/src/teamsActivityHandler.ts:669]: Could you add an empty line at the end of the file?

File: libraries/botbuilder/tests/teamsActivityHandler.test.js (hunk 3 above)
Comment (denscollo) [reply libraries/botbuilder/tests/teamsActivityHandler.test.js:1569]: @stevengum  We'll add the catch clauses to all the rest of the tests. As soon as we have it ready, we'll let you know

**Comment by denscollo** [comment]
@stevengum feedback applied. Thanks!

**Pull request merged**